src/plugins/discover/public/application/apps/main/discover_main_route.tsx

```diff
@@ -0,0 +1,150 @@
+import { redirectWhenMissing } from '../../../../../kibana_utils/public';
+import { getUrlTracker } from '../../../kibana_services';
+
+const DiscoverMainAppMemoized = memo(DiscoverMainApp);
+
+export interface DiscoverMainProps {
+  /**
+   * Instance of browser history
+   */
+  history: History;
+  /**
+   * Kibana core services used by discover
+   */
+  services: DiscoverServices;
+}
+
+interface DiscoverLandingParams {
+  id: string;
+}
+
+export function DiscoverMainRoute({ services, history }: DiscoverMainProps) {
+  const {
+    core,
+    chrome,
+    uiSettings: config,
+    data,
+    toastNotifications,
+    http: { basePath },
+  } = services;
+
+  const [savedSearch, setSavedSearch] = useState<SavedSearch>();
+  const indexPattern = savedSearch?.searchSource?.getField('index');
+  const [indexPatternList, setIndexPatternList] = useState<
+    Array<SavedObject<IndexPatternAttributes>>
+  >([]);
+
+  const { id } = useParams<DiscoverLandingParams>();
+
+  useEffect(() => {
+    const savedSearchId = id;
+
+    async function loadDefaultOrCurrentIndexPattern(usedSavedSearch: SavedSearch) {
+      await data.indexPatterns.ensureDefaultIndexPattern();
+      const { appStateContainer } = getState({ history, uiSettings: config });
+      const { index } = appStateContainer.getState();
+      const ip = await loadIndexPattern(index || '', data.indexPatterns, config);
+      const ipList = ip.list as Array<SavedObject<IndexPatternAttributes>>;
+      const indexPatternData = await resolveIndexPattern(
+        ip,
+        usedSavedSearch.searchSource,
+        toastNotifications
+      );
+      setIndexPatternList(ipList);
+      return indexPatternData;
+    }
+
+    async function loadSavedSearch() {
+      try {
+        // force a refresh if a given saved search without id was saved
+        setSavedSearch(undefined);
+        const loadedSavedSearch = await services.getSavedSearchById(savedSearchId);
+        const loadedIndexPattern = await loadDefaultOrCurrentIndexPattern(loadedSavedSearch);
+        if (loadedSavedSearch && !loadedSavedSearch?.searchSource.getField('index')) {
+          loadedSavedSearch.searchSource.setField('index', loadedIndexPattern);
+        }
+        setSavedSearch(loadedSavedSearch);
+        if (savedSearchId) {
+          chrome.recentlyAccessed.add(
+            ((loadedSavedSearch as unknown) as SavedObjectDeprecated).getFullPath(),
+            loadedSavedSearch.title,
+            loadedSavedSearch.id
+          );
+        }
+      } catch (e) {
+        redirectWhenMissing({
+          history,
+          navigateToApp: core.application.navigateToApp,
+          basePath,
+          mapping: {
+            search: '/',
+            'index-pattern': {
+              app: 'management',
+              path: `kibana/objects/savedSearches/${id}`,
+            },
+          },
+          toastNotifications,
+          onBeforeRedirect() {
+            getUrlTracker().setTrackedUrl('/');
```

Comment:
Does it somehow tested in functionals?

Reply:
this was the existing logic. discover is pretty well covered with functional tests.